modules/gapi/include/opencv2/gapi/s11n.hpp

```diff
@@ -300,14 +306,48 @@ static cv::util::optional<GCompileArg> exec(const std::string& tag, cv::gapi::s1
 
 template<typename T> struct deserialize_runarg;
 
-template<typename RMatAdapterType>
-struct deserialize_runarg {
+template<typename T, typename... Types> inline
+typename std::enable_if<std::is_base_of<RMat::Adapter, T>::value, GRunArg>::
+type deserializeRMatHelper(cv::gapi::s11n::IIStream& is, uint32_t) {
+    auto ptr = std::make_shared<T>();
+    ptr->deserialize(is);
+    return GRunArg { RMat(std::move(ptr)) };
+}
+
+template<typename T, typename... Types> inline
+typename std::enable_if<std::is_base_of<MediaFrame::IAdapter, T>::value, GRunArg>::
+type deserializeRMatHelper(cv::gapi::s11n::IIStream& is, uint32_t idx) {
+    return deserialize_runarg<std::tuple<Types...>>::exec(is, idx);
+}
+
+template<typename T, typename... Types> inline
+typename std::enable_if<std::is_base_of<MediaFrame::IAdapter, T>::value, GRunArg>::
+type deserializeMediaFrameHelper(cv::gapi::s11n::IIStream& is, uint32_t) {
+    std::unique_ptr<T> ptr(new T);
+    ptr->deserialize(is);
+    return GRunArg { MediaFrame(std::move(ptr)) };
+}
+
+template<typename T, typename... Types> inline
+typename std::enable_if<std::is_base_of<RMat::Adapter, T>::value, GRunArg>::
+type deserializeMediaFrameHelper(cv::gapi::s11n::IIStream& is, uint32_t idx) {
+    return deserialize_runarg<std::tuple<Types...>>::exec(is, idx);
+}
+
+template<> struct deserialize_runarg<std::tuple<>> {
+static GRunArg exec(cv::gapi::s11n::IIStream&, uint32_t) {
+        GAPI_Assert(false && "Failed to deserialize MediaFrame or RMat - mismatched adapter types");
+    }
+};
+
+template<typename T, typename... Types>
+struct deserialize_runarg<std::tuple<T, Types...>> {
 static GRunArg exec(cv::gapi::s11n::IIStream& is, uint32_t idx) {
     if (idx == GRunArg::index_of<RMat>()) {
-        auto ptr = std::make_shared<RMatAdapterType>();
-        ptr->deserialize(is);
-        return GRunArg { RMat(std::move(ptr)) };
-    } else { // non-RMat arg - use default deserialization
+        return deserializeRMatHelper<T, Types...>(is, idx);
```

Comment:
Note previously the code constructed an object right here using the type `RMatAdapterType`.

You can still do it like that:

```c++
    if (idx == GRunArg::index_of<RMat>()) {
        using RMatAdapterType = util::find_adapter_impl<RMat::Adapter, Types...>::type;
        auto ptr = std::make_shared<RMatAdapterType>();
        ptr->deserialize(is);
        return GRunArg { RMat(std::move(ptr)) };
    } else { // non-RMat arg - use default deserialization
```
All you need to do is to introduce a new helper `util::find_adapter_impl` which identifies which type of `Types...` implements a certain interface - and no need to spread the logic in four more functions as in above

Reply:
I don't think it's possible here. Please, correct me if I'm wrong. Example:
`find_adapter_impl`:
```
    template<typename T>
    struct find_adapter_impl;

    template<typename AdapterT, typename T>
    struct find_adapter_impl<std::tuple<AdapterT, T>>
    {
        using type = typename std::conditional<std::is_base_of<AdapterT, T>::value,
                                               T,
                                               void>::type;
        static constexpr bool found = std::is_base_of<AdapterT, T>::value;
    };

    template<typename AdapterT, typename T, typename... Types>
    struct find_adapter_impl<std::tuple<AdapterT, T, Types...>>
    {
        using type = typename std::conditional<std::is_base_of<AdapterT, T>::value,
                                               T,
                                               typename find_adapter_impl<AdapterT, Types...>::type>::type;
        static constexpr bool found = std::is_base_of<AdapterT, T>::value ||
                                      find_adapter_impl<AdapterT, Types...>::found;
    };
```
Usage:
```
if (idx == GRunArg::index_of<RMat>()) {
        GAPI_Assert((cv::detail::find_adapter_impl<std::tuple<RMat::Adapter, T, Types...>>::found) &&
                    "Can't find an adapter type in the parameters to deserialize RMat");
        using RMatAdapterType =
                typename cv::detail::find_adapter_impl<std::tuple<RMat::Adapter, T, Types...>>::type;
        auto ptr = std::make_shared<RMatAdapterType>();
        ptr->deserialize(is);
        return GRunArg { RMat(std::move(ptr)) };
    } else if (idx == GRunArg::index_of<MediaFrame>()) {
        GAPI_Assert((cv::detail::find_adapter_impl<std::tuple<MediaFrame::IAdapter, T, Types...>>::found) &&
                    "Can't find an adapter type in the parameters to deserialize MediaFrame");
        using MediaFrameAdapterType =
                typename cv::detail::find_adapter_impl<std::tuple<MediaFrame::IAdapter, T, Types...>>::type;
        std::unique_ptr<MediaFrameAdapterType> ptr(new MediaFrameAdapterType);
        ptr->deserialize(is);
        return GRunArg { MediaFrame(std::move(ptr)) };
    }
```

Note: during compile time it will compile both cases: `RMat` and `MediaFrame`. But what if we get only one of them?
Example:
`cv::gapi::deserialize<cv::GRunArgs, MyRMatAdapter>(sargsin);`
It will fail the compilation of `MediaFrame` part since it's impossible to get the proper type there (for `new`, `->`, etc)

Reply:
It's not `constexpr if` so we can't prevent the compilation for some branches. For c++11 `enable_if` is the only solution I know

Reply:
Alternatively, we can try to change `void` type to `std::conditional` for the base adapter type depending on the input template, but it would look rather ugly

Reply:
I don't see any limitations here. This is how the below code should look like:

```c++
    if (idx == GRunArg::index_of<RMat>()) {
        // Type or void (if not found)
        using TA = cv::detail::find_adapter_impl<RMat::Adapter, Types...>>::type;
        return deserialize_construct<RMat, TA>(is);
    } else if (idx == GRunArg::index_of<MediaFrame>()) {
        // Type or void (if not found)
        using TA = cv::detail::find_adapter_impl<MediaFrame::IAdapter, Types...>>::type;
        return deserialize_construct<MediaFrame, TA>(is);
    }
```

where

```c++
template<typename G, typename TA>
GRunArg deserialize_construct(IIStream &is) {
    std::unique_ptr<TA> ptr(new TA{});
    ptr->deserialize(is);
    return GRunArg { G(std::move(ptr)) };
}


template<typename G>
GRunArg deserialize_construct<G,void>(IIStream &is) {
    Assert("No suitable adapter class found");
}
```

Also: please don't introduce a need in extra `std::tuple` to handle `T...` -- in many cases it is just not required.

Reply:
Done. Removed `tuple` where possible. Also, used `struct`s instead of plain function since those can be partially specialized